sidechain/src/lib.rs

```diff
@@ -110,47 +108,41 @@ pub mod pallet {
 				return Ok(().into())
 			}
 
-			let lenience = T::EarlyBlockProposalLenience::get();
-			let mut latest_confirmation = <LatestSidechainBlockConfirmation<T>>::get(shard_id);
+			let latest_confirmation = <LatestSidechainBlockConfirmation<T>>::get(shard_id);
 			let latest_block_number = latest_confirmation.block_number;
 			let block_number = confirmation.block_number;
+			let finalization_candidate_block_number =
+				<SidechainBlockFinalizationCandidate<T>>::try_get(shard_id).unwrap_or(1);
+
+			<SidechainBlockFinalizationCandidate<T>>::insert(
+				shard_id,
+				next_finalization_candidate_block_number,
+			);
 
-			if block_number > Self::add_to_block_number(latest_block_number, lenience)? {
-				// Block is far too early and hence refused.
-				return Err(<Error<T>>::BlockNumberTooHigh.into())
-			} else if block_number > Self::add_to_block_number(latest_block_number, 1)? {
-				// Block is too early and stored in the queue for later import.
-				if !<SidechainBlockConfirmationQueue<T>>::contains_key((shard_id, block_number)) {
-					<SidechainBlockConfirmationQueue<T>>::insert(
-						(shard_id, block_number),
-						confirmation,
-					);
-				}
-			} else if block_number == Self::add_to_block_number(latest_block_number, 1)? {
+			if block_number > finalization_candidate_block_number {
+				// Block is too early and hence refused.
+				Err(<Error<T>>::BlockNumberTooHigh.into())
+			} else if block_number == finalization_candidate_block_number {
 				Self::finalize_block(shard_id, confirmation, &sender, sender_index);
-				latest_confirmation = confirmation;
-
-				Self::finalize_blocks_from_queue(
-					shard_id,
-					&sender,
-					sender_index,
-					latest_confirmation,
-				)?;
+				Ok(().into())
+			} else if block_number > latest_block_number {
+				// Block does not belong to the finalized ones.
+				Err(<Error<T>>::ReceivedUnexpectedSidechainBlock.into())
 			} else {
 				// Block is too late and hence refused.
-				return Err(<Error<T>>::OutdatedBlockNumber.into())
+				Err(<Error<T>>::OutdatedBlockNumber.into())
 			}
```

Comment:
I think the branches here are a bit confusing now. I guess we simplify the code here to:

```rust

if block_number == finalization_candidate_block_number {
    Self::finalize_block(...)
    /// see comment above
   <SidechainBlockFinalizationCandidate<T>>::insert(
				shard_id,
				next_finalization_candidate_block_number,
   );
} else {
    Err(<Error<T>>::ReceivedUnexpectedSidechainBlock.into())
}
```

Reply:
This works even better or more conventional would be to use ensure.

```rust
// Check error condition
ensure!(block_number == finalization_candidate_block_number, Error::<T>::ReceivedUnexpectedSidechainBlock);
// Now you are in your valid state now mutate storage.
Ok(().into())
```

Reply:
nice, i agree.